File: python-starter/src/benchmark/observation.py
"""Page observation capture for the benchmark framework."""

from __future__ import annotations

import os

from playwright.async_api import Page

from .browser_adapter import BrowserAdapter
from .types import Observation


async def capture_observation(
    page: Page,
    adapter: BrowserAdapter,
    *,
    screenshot_dir: str,
    step_id: str,
) -> Observation:
    """Capture the current page state including screenshot and DOM evidence."""
    os.makedirs(screenshot_dir, exist_ok=True)
    screenshot_path = os.path.join(screenshot_dir, f"{step_id}.png")
    await adapter.screenshot(screenshot_path)

    url = page.url
    title = await page.title()

    headings = await page.locator("h1, h2").evaluate_all(
        """elements => elements
            .map(el => (el.textContent || '').trim())
            .filter(Boolean)"""
    )

    banners = await page.locator(
        "[role='alert'], [data-status-banner]"
    ).evaluate_all(
        """elements => elements
            .map(el => (el.textContent || '').trim())
            .filter(Boolean)"""
    )

    dom_evidence = ""
    try:
        raw = await page.locator("main").inner_text()
        dom_evidence = " ".join(raw.split())[:500]
    except Exception:
        pass

    controls = await adapter.extract_controls()

    return Observation(
        url=url,
        title=title,
        headings=headings,
        banners=banners,
        dom_evidence=dom_evidence,
        screenshot_path=screenshot_path,
        controls=controls,
    )

File: starter/src/benchmark/observation.js
import { mkdir } from "node:fs/promises";
import path from "node:path";

export async function captureObservation(page, adapter, { screenshotDir, stepId }) {
  await mkdir(screenshotDir, { recursive: true });
  const screenshotPath = path.join(screenshotDir, `${stepId}.png`);
  await adapter.screenshot(screenshotPath);

  const url = page.url();
  const title = await page.title();
  const headings = await page
    .locator("h1, h2")
    .evaluateAll((elements) => elements.map((element) => element.textContent?.trim() || "").filter(Boolean));
  const banners = await page
    .locator("[role='alert'], [data-status-banner]")
    .evaluateAll((elements) => elements.map((element) => element.textContent?.trim() || "").filter(Boolean));
  const domEvidence = await page
    .locator("main")
    .innerText()
    .then((text) => text.replace(/\s+/g, " ").trim().slice(0, 500))
    .catch(() => "");
  const controls = await adapter.extractControls();

  return {
    url,
    title,
    headings,
    banners,
    domEvidence,
    screenshotPath,
    controls
  };
}
